merge common.diag and dense.diag

smichr opened the pull request:
#### Brief description of what is fixed or changed
- Matrix([1, [2]]) now raises an error
- diag shape defaults to square if only rows or cols is given
- keywords allow control of Matrix.diag since dense.diag behaved different than common.diag and the two are now merged.

The publicly facing `diag` and private `Matrix.diag` now have controls:

1. a single list can be unpacked instead of being converted to a matrix with keyword `unpack`
```python
>>> diag([1, 2, 3])
Matrix([
[1],
[2],
[3]])
>>> diag([1, 2, 3], unpack=True) # = diag(*[1, 2, 3]) = diag(1, 2, 3) = Matrix.diag([1, 2, 3])
Matrix([
[1, 0, 0],
[0, 2, 0],
[0, 0, 3]]
```

2. The conversion of lists to matrices can be relaxed so list of ragged rows can be accepted:
```python
>>> diag([[1, 2, 3], 4], strict=False)
Matrix([
[1, 2, 3],
[4, 0, 0]])
```
This is the default behavior of Matrix.diag (but unpack must be False for such a single list or an additional level of brackets must be added)
```python
>>> Matrix.diag([[1, 2, 3], 4], unpack=False)
Matrix([
[1, 2, 3],
[4, 0, 0]])
>>> Matrix.diag([  [[1, 2, 3], 4]  ])
Matrix([
[1, 2, 3],
[4, 0, 0]])
```

(Such input, formerly, put list-literals into the resulting matrix, making it non-functional.)

#### Other comments


During previous cleanup of the matrix files structure, two versions
of `diag` were left behind. The one that is public facing via
`from sympy import *` is the one in `dense.py`. It differs from
that in `common.py` a.k.a `Matrix.diag` by converting an input list to a matrix
rather than unpacking it. Hence, it is strict about matrix structure.

The `Matrix.diag` was not strict and actually could produce non-functional
matrices. An aspect of non-strictness is retained while doing the right thing
more often; an option to strictly convert lists to matrices is controlled with
the keyword `strict`. It is True for the public `diag` and False for the private `Matrix.diag`.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- matrices
    - matrix creation detects singleton lists in matrices so Matrix([1, [2]]) now raises an error
    - diag matrix shape defaults to square if only rows or cols is given
    - diag unpacking of lists (conversion of single lists to matrices) can be controlled with keyword `unpack`; dense.diag defaults to True while Matrix.diag (common.diag) defaults to False; `Matrix.diag(*[1, 2, 3]) == Matrix.diag([1, 2, 3], unpack=True)`)
    - diag accepts variable-length rows if keyword `strict` is set to False
<!-- END RELEASE NOTES -->

sylee957 commented:
The merging would look good.

But I have some questions about this implementation.

<img src="https://user-images.githubusercontent.com/34944973/54664149-c35cf380-4b26-11e9-9305-49d175d8bf98.png" width="256px">

From first and second examples, I find that this change is backward incompatible.

And from the third one, is it valid?

smichr commented:
I pushed a commit that retains the unpacking difference between the two. I think it would be good to deprecate this difference, however.

I'm not sure what to do about the "fill diagonal" vs "fill from last diagonal". Like I said, I think this is an oversight and an undesirable characteristic that would only affect filling with non-square elements.

smichr commented:
@asmeurer , could you give a thumbs up or down with making the following change without deprecation? I think the previous behavior does not make sense -- is more a bug than a feature -- for a method named "diag" which should be filling items in on the diagonal:
```python
>>> diag([1, 2, 3], [4, 5], 6)
Matrix([
[1, 0, 0],
[2, 4, 0],
[3, 5, 6]])

FORMERLY

Matrix([
[1, 0, 0],
[2, 0, 0],
[3, 0, 0],
[0, 4, 0],
[0, 5, 0],
[0, 0, 6]])
```

smichr commented:
I've got to think about how to combine these better.

smichr commented on sympy/matrices/dense.py:
```diff
@@ -1103,103 +1103,42 @@ def eye(*args, **kwargs):
+        kwargs.pop('unpack')
+        return rv
+    else:
+        return Matrix.diag(*values, **kwargs)
```
Oh, that's the way to add an additional kword. Thanks!

jksuom replied:
That will not work as long as Python 2.7 is supported.

asmeurer replied:
I think it works. The problem in 2.7 is doing that in the parameter list of a function definition.

smichr committed:
inline edit

smichr committed:
detect singleton lists with ints

smichr committed:
allow control of Matrix.diag from dense.diag

dense.diag is the public facing diag routine that
never got merged with common.diag (accessible through Matrix.diag)
and each routine has different semantics when it comes to
unpacking and treatment of arguments that are lists.

Matrix.diag unpacked singleton sequences and did not convert other
lists to matrices

dense.diag converted all sequences to matrices and hence did not
unpack args

The behavior of dense.diag -- the publicly facing routine imported
with 'from sympy import *' is unchanged.

The behavior of common.diag a.k.a Matrix.diag has been modified to
treat lists as matrices rather than to try place list-literals
into a matrix (which creates a non-functional matrix):

>>> Matrix.diag([1,2],[3])
Matrix([
[[1, 2],   0],
[     0, [3]]])

The "wandering diagonal" behavior is retained so a
block-diagonal can be created from non-square matrices.
If the elements always start filling on the main diagonal,
blocks will end up overlapping. So the folling is correct:

>>> diag([1,2,3],4)
Matrix([
[1, 0],
[2, 0],
[3, 0],
[0, 4]])

This is not a block-diagonal since the "4" block overlaps
with the first column.

>>> diag([1,2,3],4)
Matrix([
[1, 0],
[2, 4],
[3, 0]])

asmeurer commented:
Is there a motivation for the breaking change? I'm not really sure which behavior would be more useful.  

The equivalent I found for NumPy is `scipy.linalg.block_diag`, which works like `sympy.diag` does in master https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.linalg.block_diag.html. 

Regarding the history of this, there was originally `block_diag`, which only supported square matrices. Then `diag` was added in 25212c818dfaaef8db4eb43bb951dd4b91afa62c, which supported any shape (the behavior in master). Eventually `block_diag` was deprecated and removed in favor of `diag`.

asmeurer commented:
The current behavior of diag is how I would read a block matrix with non-square blocks on the diagonal. Like https://en.wikipedia.org/wiki/Block_matrix#Block_diagonal_matrices if the A1, A2, ... were not square. 

I think we should keep the current `diag` behavior. I think it is useful (at least since it is how scipy's block_diag works), and it has been there for a while. We can add the other behavior under a flag, although I'm still unclear when exactly you would use it.

smichr commented:
Very helpful history, @asmeurer. Thanks. I missed the deprecation of block_diag. It makes sense now to keep the old behavior. A more general `banded` function in progress will be able to create matrices which are described by the elements on the diagonals.

I also see how the notion of block would make the fill-from-last- diagonal make sense...if you don't then you aren't describing blocks.

smichr commented:
Commit history will be cleaned up and I will make any additional changes that are requested. @sylee957 , are you able to give this a look again? There is now no backwards incompatible changes.

smichr commented:
- [ ] note that diag of non-square matrices is just the "direct sum" referenced [here](https://en.wikipedia.org/wiki/Block_matrix)

smichr committed:
Update common.py
